Overhaul the rust thrift library

reem opened the pull request:
The major changes are:
    - Concrete impls of Transport have been removed in favor of a blanket implementation.
    - Implement standard error handling idioms, including implementations of `std::error::Error` for all error types.
    - Various renames to bring things in line with idiomatic rust.
    - Change the Protocol trait to use static calls rather than virtual dispatch.
    - Allow Protocol's to mutate an internal state by changing the trait methods to receive &mut self.
    - Processor::process now receives &self ahead of a multithreaded server implementation.

This can't be landed as-is since there need to be accompanying changes to the code generator, but I wanted to solicit feedback on these changes first. I also have a multithreaded server ready to be PRed based on this branch, which you can see [here](https://github.com/terminalcloud/thrift/blob/threaded-server/lib/rs/src/server/threaded.rs).

reem committed:
Add a blanket implementation for Transport of all T: Read + Write

Also remove the now unnecessary dependency on the bufstream crate,
since the implementation it is used in is now included in the blanket
implementation of Transport.

reem committed:
Implement std Error and appropriate From conversions for errors

Allows for cause chains containing high-level thrift errors,
protocol errors, and decoding errors.

reem committed:
Overhaul the Protocol trait to use static calls

Removes all use of virtual dispatch on transports and protocols, allowing
for enhanced inlining and other optimization support.

Also renames the Type and MessageType variants to not include a T prefix.

reem committed:
Rename ServerTransport and update simple_server for changes.

reem committed:
Allow Protocol's to mutate an internal state.

reem committed:
Rename protocol::{Read, Write} to protocol::{Encode, Decode}.

reem committed:
Change Processor::process to receive &self

This allows a Sync Processor to be used in a multithreaded context.

gsingh93 commented:
Apparently Github only shows the context for comments on the pull request itself, not the actual commits. That's annoying :/

Thanks for the pull request! Overall these changes look great to me, we'll wait for @maximg to chime in with any other comments.

I'm not sure about the Encode/Decode rename. While Read/Write aren't ideal either, they at least express that data is being written (and possibly sent) on the transport, not just encoded/decoded and waiting to be sent.

@maximg Can you push the benchmark you were using so we can see how changes like these affect performance?

maximg commented:
Thanks for the PR! Will go over it this evening. Just committed the benchmark so feel free to compare.

reem commented:
I am looking to make some changes to the code generator to merge these changes as well as add support for more of thrift, notably exceptions. However, I am a novice C++ programmer so it's taking me some time to become familiar with the generator - if either of you has a minute I'd love to chat a little about the code to help me get productive faster.

gsingh93 commented:
Feel free to ping me on IRC.

maximg commented:
For exceptions could you make sure to check with @gsingh93 - he is working on them already. I'll be happy to chat on skype (maxim.golov)

reem committed:
Implement a BufferedTransportServer wrapper.

Forgetting to buffer your Transport streams is a major performance gotcha,
so we provide BufferedTransportServer to wrap any existing TransportServer,
e.g. TcpListener, to generate buffered streams over the underlying Transport.

reem commented:
I've pushed another commit adding a `BufferedTransportServer` wrapper which causes the yielded `Transport`s to be wrapped in a `BufStream`. I think it makes sense to provide this as part of the library since it's a major performance gotcha.

reem commented:
We _may_ even want to integrate `BufStream` into the impl of `TransportServer` for `TcpListener` so the simple thing (passing a `TcpListener` as your `TransportServer`) does the right thing.

gsingh93 commented:
The buffered transport looks good, but I don't think it's necessary to add it to the `TcpListener` impl. We're going to eventually follow the same pattern as the other libs and uses a factory ([example](https://github.com/maximg/thrift/blob/dev-rust-thrift/tutorial/py/PythonServer.py#L86)), which will require the user to pass in a factory instead of a transport. The most basic factory will be a buffered factory, so the user shouldn't run into any surprises with their stream not buffering.

Also, I think `TransportServer` should be changed backed to `ServerTransport`. This is what the other libraries use, as well as the [original thrift whitepaper](https://thrift.apache.org/static/files/thrift-20070401.pdf). It does make sense, as it's the transport the server uses to talk to the client, it isn't the server itself (although both interpretations somewhat make sense).

reem commented:
Ok, I can change it back.
